src/include/miopen/inline_vector.hpp

```diff
@@ -0,0 +1,251 @@
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifndef GUARD_MIOPEN_INLINE_VECTOR_HPP
+#define GUARD_MIOPEN_INLINE_VECTOR_HPP
+
+#include <array>
+#include <miopen/config.h>
+#include <miopen/errors.hpp>
+
+namespace miopen {
+
+template <typename T, std::size_t N>
+class InlineVector
+{
+public:
+    using value_type = T;
+    static_assert(std::is_scalar_v<T>, "Input data size is bigger than InlineVector's capacity");
+
+    // Default constructor
+    InlineVector() = default;
+
+    // Copy and move constructor
+    InlineVector(const InlineVector& inline_vec)     = default;
+    InlineVector(InlineVector&& inline_vec) noexcept = default;
+
+    InlineVector(std::initializer_list<T> __data) : real_size(__data.size())
+    {
+        if(real_size > N)
+        {
+            MIOPEN_THROW("Input data size is bigger than InlineVector's capacity");
+        }
+
+        std::copy(__data.begin(), __data.end(), _data.begin());
+    }
+
+    template <typename _InputIterator, typename = std::_RequireInputIter<_InputIterator>>
+    InlineVector(_InputIterator first, _InputIterator last) : real_size(std::distance(first, last))
+    {
+        if(real_size > N)
+        {
+            MIOPEN_THROW("Input data size is bigger than InlineVector's capacity");
+        }
+
+        std::copy(first, last, _data.begin());
+    }
+
+    // Copy/move operator
+    InlineVector& operator=(const InlineVector& inline_vec)     = default;
+    InlineVector& operator=(InlineVector&& inline_vec) noexcept = default;
+
+    // Iterators
+    T* begin() noexcept { return _data.begin(); }
+
+    const T* begin() const noexcept { return _data.begin(); }
+
+    T* end() noexcept { return (_data.begin() + real_size); }
+
+    const T* end() const noexcept { return (_data.begin() + real_size); }
+
+    // Constant iterator
+    const T* cbegin() const noexcept { return begin(); }
+
+    const T* cend() const noexcept { return end(); }
```

Comment:
MSVC is strict about iterators types, please replace `T*` with proper iterators.
You can add typedefs for pointer/reference/iterator and reuse them from `std::array`.
```c++
using storage_type = std::array<T, N>;
using iterator = storage_type::iterator;
...etc...
private:
storage_type _data;
```